### PR TITLE
perf(core): snapshots

### LINKED ISSUE #1997: perf: state snapshots
Having to process the state trie to query all account information, as well as having to compute the sender's public key from its address at least once per transaction, introduces significant overhead to both L1 and L2.
This must be done when receiving new data from other nodes, but it's not necessary for internal operation, given we either already verified the information or created it ourselves. We could use a flat representation of current state most of the time.
The observation is not new, and [Geth already uses state snapshots](https://blog.ethereum.org/2020/07/17/ask-about-geth-snapshot-acceleration).

While this would introduce significant complexity, it would also make a big reduction in the time we take to include transactions in the mempool, according to profiles performed on load tests for the L2.

In terms of design, we could start with something simpler than Geth, just a table on the DB with each address's data on the latest state, recomputing from the trie on reorgs (e.g. we could have a generation number of some sort to detect stale data).

### PR DESCRIPTION
**Pull request by pablodeymo**
# Account Info and Account Storage Snapshot Feature
## Goal
Improve database read access performance for block execution, while keeping write access overhead in check.

## MVP Requirements

- Ability to build a snapshot from a point in time and keep it up-to-date.
- Ability to handle re-orgs correctly.
- Ability to read from the snapshot when its state corresponds to the one for the base block.

## Missing for MVP

- Check for re-orgs beyond start of log. It should discard the snapshot and start over.
- Implementations for InMemoryDB and ReDB.
- Unit or integration tests for edge cases.
- Simplify by using smaller helpers inside the DB implementations. They need to take either the open transaction or the open cursor to operate on, to avoid adding commit overhead.

## V2 Requirements

- Unify tables with snap sync.
- Fix the healing process of snap sync to produce a re-usable snapshot.
- Implementation of log pruning, to limit the size of the snapshot.
- Work directly on `AccountUpdate`s for simplicity.

## Later Work

- The choice of using a log was made to enable a later change of index for tries, using their prefix rather than their hashes as key.
- Under that implementation, the snapshot becomes one with the trie, as finding leaves is finding values and their prefix is the whole path in the trie.
- There might be some extra improvement both in code size and storage size in saving the `XOR` of old and new values instead of the values themselves in the log entries. The storage should become about half, while the code for rewinding and replaying would change only in the direction of iteration, as `XOR` is a symmetric operation.

closes #1997

### CODE REVIEW

File: crates/storage/store.rs [hunk 1]
```diff
@@ -108,6 +138,88 @@ impl Store {
...
+                        .engine
+                        .get_current_account_info(address)?
+                        .unwrap_or_default(),
+                };
+                // NOTE: this might also be useful as preparation for the snapshot
+                previous_account_info.insert(address, new_info.clone());
+                accounts_info_log.push((block_numhash, address.into(), old_info, new_info.clone()));
+            }
+        }
+        Ok(accounts_info_log)
```
Comment (Oppen) [inline crates/storage/store.rs]: Missed edge-case: `account_update.removed`.

File: crates/storage/store_db/redb.rs [hunk 2]
```diff
@@ -301,29 +320,418 @@ impl RedBStore {
             .transpose()
             .map_err(StoreError::from)
     }
+
+    /// Process account info logs and revert changes to the account_info_table
```
[the remark this answers is not in the document]
Comment (damiramirez) [reply crates/storage/store_db/redb.rs:324]: Will do this in another PR

File: crates/storage/store_db/redb.rs [hunk 3]
```diff
@@ -301,29 +320,418 @@ impl RedBStore {
...
+
     async fn apply_updates(&self, update_batch: UpdateBatch) -> Result<(), StoreError> {
         let db = self.db.clone();
         tokio::task::spawn_blocking(move || {
             let write_txn = db.begin_write().map_err(Box::new)?;
             {
-                // store account updates
+                // We only need to update the flat tables if the update batch contains blocks
+                // We should review what to do in a reconstruct scenario, do we need to update the snapshot state?
+                if let (Some(first_block), Some(last_block)) = (update_batch.blocks.first(), update_batch.blocks.last()) {
```
[the remark this answers is not in the document]
Comment (damiramirez) [reply crates/storage/store_db/redb.rs:651]: Discussing with the team

File: crates/storage/store_db/in_memory.rs [hunk 4]
```diff
@@ -147,6 +233,212 @@ impl StoreEngine for Store {
...
+                        store.account_storage.remove(&(log.address, log.slot));
+                    } else {
+                        store
+                            .account_storage
+                            .insert((log.address, log.slot), log.old_value);
+                    }
+
+                    // We update this here to ensure it's the previous block according
+                    // to the logs found.
+                    BlockNumHash(block_num, snapshot_hash) = parent_block;
```
[the remark this answers is not in the document]
Comment (damiramirez) [reply crates/storage/store_db/in_memory.rs]: Both variables are mutable
```
let mut block_num = current_snapshot.0;
let mut snapshot_hash = current_snapshot.1;
```

File: crates/storage/api.rs [hunk 5]
```diff
@@ -307,6 +319,32 @@ pub trait StoreEngine: Debug + Send + Sync + RefUnwindSafe {
...
+
+    /// Sets up the flat tables of account info for the genesis block
+    async fn setup_genesis_flat_account_info(
+        &self,
+        genesis_block_number: u64,
+        genesis_block_hash: H256,
+        genesis_accounts: &[(Address, u64, U256, H256, bool)],
+    ) -> Result<(), StoreError>;
+
+    /// Gets the block hash for the current snapshot
```
Comment (Arkenan) [inline crates/storage/api.rs:338]: nit: maybe clarify that the result is an option because there may be no current snapshot (if there is, there will always be an associated block hash).

File: crates/storage/error.rs [hunk 6]
```diff
@@ -53,3 +53,10 @@ pub enum StoreError {
     #[error("Failed to lock database for writing")]
     LockError,
 }
+
+#[cfg(feature = "libmdbx")]
+impl From<anyhow::Error> for StoreError {
+    fn from(err: anyhow::Error) -> Self {
+        StoreError::LibmdbxError(err)
+    }
+}
```
Comment (Arkenan) [inline crates/storage/error.rs:62]: This seems convoluted and counterintuitive. anyhow:Error is very generic. How can we make sure it would always be a libmdbx error? why do we need this?

File: crates/storage/rlp.rs [hunk 7]
```diff
@@ -24,6 +31,22 @@ pub type AccountHashRLP = Rlp<H256>;
...
+#[cfg(feature = "redb")]
+pub type BlockNumHashRLP = Rlp<BlockNumHash>;
+#[cfg(feature = "redb")]
+pub type AccountAddressRLP = Rlp<Address>;
+#[cfg(feature = "redb")]
+pub type AccountInfoRLP = Rlp<AccountInfo>;
+#[cfg(feature = "redb")]
+pub type AccountStorageKeyRLP = Rlp<H256>;
+#[cfg(feature = "redb")]
+pub type AccountStorageValueRLP = Rlp<U256>;
```
Comment (Arkenan) [inline crates/storage/rlp.rs:48]: If this is redb specific, why do we need it in the storage rlp directory? Can we not have it in the redb file instead?

File: crates/storage/store.rs [hunk 8]
```diff
@@ -108,6 +114,99 @@ impl Store {
...
+    // 1. From the number of the first new_canonical_block and the canonical
+    //    chain table, extract the list of (block_number, block_hash) we need
+    //    to invalidate
+    // 2. From the last block in that list, iterate backwards the log restoring
+    //    the previous values
+    // 3. From the first block in the new canonical chain to the last, apply
+    //    the log for those blocks
+    // 4. Commit the transaction
+    //
+    // This function assumes that `new_canonical_blocks` is sorted by block number.
```
Comment (Arkenan) [inline crates/storage/store.rs:127]: What is "new canonical blocks"?

File: crates/storage/store.rs [hunk 9]
```diff
@@ -108,6 +114,99 @@ impl Store {
...
+        &self,
+        head: H256,
+    ) -> Result<(), StoreError> {
+        self.engine.undo_writes_until_canonical().await?;
+        self.engine.replay_writes_until_head(head).await
+    }
+
+    // TODO: build_account_info_logs and build_account_storage_logs became a bit
+    // redundant. Just pass the account_updates whole and process inside the
+    // apply_updates function.
```
Comment (Arkenan) [inline crates/storage/store.rs:138]: This sounds very sensible. Let's add an issue for this.

File: crates/storage/store_db/libmdbx.rs [hunk 10]
```diff
@@ -122,34 +281,355 @@ impl Store {
...
+        &self,
+        from_snapshot: BlockNumHash,
+        to_snapshot: BlockNumHash,
+        state_log_cursor: &mut libmdbx::orm::Cursor<'_, libmdbx::RW, AccountsStateWriteLog>,
+        flat_info_cursor: &mut libmdbx::orm::Cursor<'_, libmdbx::RW, FlatAccountInfo>,
+    ) -> Result<(), StoreError> {
+        let mut state_logs = state_log_cursor.seek_closest(to_snapshot)?;
+
+        let mut account_updates = vec![];
+        while let Some(((final_block, parent_block), log_entry)) = state_logs {
```
Comment (Arkenan) [inline crates/storage/store_db/libmdbx.rs:462]: nit: I see we repeat this pattern quite a lot (seek closest + while). Shall we refactor it to an iterator se we can use it like this? `self.get_logs(cursor, snapshot).map(...)`.

Not for this PR and also optional, might be more complex than it seems.

**Review by Arkenan — APPROVED** [review]
LGTM, all of my comments can be addressed in a separate PR.

**Review by jrchatruc — CHANGES_REQUESTED** [review]
Blocking merge of this PR until we run the numbers along with pruning

**Comment by mpaulucci** [comment]
Moving to draft

**Comment by damiramirez** [comment]
I think we can close this PR since it didn't give us the results we expected. What do you think, @Oppen?
